File: polyglot/Bridge/bridge.hs
-- Bridge design patter

-- bridge implementation interface
class IDrawAPI a where
  drawCircle :: a -> Int -> Int -> Int -> IO ()

-- concrete bridge implementation classes implementing DrawAPI interface
data RedCircle = RedCircle
instance IDrawAPI RedCircle where
  drawCircle _ radius x y =   putStr "Drawing Circle[ color: red  , radius: "
                          >>  putStr (show radius)
                          >>  putStr ", x: " >> putStr (show x)
                          >>  putStr ", y: " >> putStr (show y)
                          >>  putStrLn "]"

data GreenCircle = GreenCircle
instance IDrawAPI GreenCircle where
  drawCircle _ radius x y =   putStr "Drawing Circle[ color: green, radius: "
                          >>  putStr (show radius)
                          >>  putStr ", x: " >> putStr (show x)
                          >>  putStr ", y: " >> putStr (show y)
                          >>  putStrLn "]"

data DrawAPI = DrawAPIr RedCircle | DrawAPIg GreenCircle
instance IDrawAPI DrawAPI where
  drawCircle (DrawAPIr a) radius x y = drawCircle a radius x y
  drawCircle (DrawAPIg a) radius x y = drawCircle a radius x y

makeRedAPI = (DrawAPIr RedCircle)
makeGreenAPI = (DrawAPIg GreenCircle)

-- create an abstract class Shape using the DrawAPI interface.
class IShape a where  -- just the interface, no data
  draw :: a -> IO ()

data Shape = Shape {_drawAPI::DrawAPI}
instance IShape Shape where
  draw _ = putStrLn "Shape::draw is not implemented"

-- create concrete class implementing the Shape interface (abstract class)
data Circle = Circle {_x::Int, _y::Int, _radius::Int, _shape::Shape}
makeCircle :: Int -> Int -> Int -> DrawAPI -> Circle
makeCircle x y radius drawAPI = Circle x y radius (Shape drawAPI)
instance IShape Circle where
  draw (Circle x y radius (Shape drawAPI)) = drawCircle drawAPI radius x y

-- Use Shape and DrawAPI classes to draw different colored circles
main = do 
   -- implementation of concrete circles selected at run time.
  let redCircle = makeCircle 100 100 10 makeRedAPI;
      greenCircle = makeCircle 100 100 10 makeGreenAPI in do 
      draw redCircle
      draw greenCircle

File: poly/DesignPatterns/Bridge/bridge.cs
// Bridge design pattern
using System;

// bridge implementation interface
interface DrawAPI {
  void DrawCircle(int radius, int x, int y);
}

// concrete bridge implementation classes implementing DrawAPI interface
class RedCircle : DrawAPI {
  public void DrawCircle(int radius, int x, int y){
    Console.WriteLine(
        "Drawing Circle[ color: red  , radius: "+radius+", x: "+x+ ", y: "+y+"]"
    );
  }
}

class GreenCircle : DrawAPI {
  public void DrawCircle(int radius, int x, int y){
    Console.WriteLine(
        "Drawing Circle[ color: green, radius: "+radius+", x: "+x+ ", y: "+y+"]"
    );
  }
}

// create an abstract class Shape using the DrawAPI interface.
abstract class Shape {
  protected DrawAPI drawAPI;
  protected Shape(DrawAPI drawAPI){this.drawAPI = drawAPI;}
  public abstract void Draw();  // abstraction need not follow implementation API
}

// create concrete class implementing the Shape interface (abstract class)
class Circle : Shape {
  private int x, y, radius;
  public Circle(int x, int y, int radius, DrawAPI drawAPI) : base(drawAPI){
    this.x = x;
    this.y = y;
    this.radius = radius;
  }
  public override void Draw(){
    drawAPI.DrawCircle(radius,x,y);
  }
}

// Use Shape and DrawAPI classes to draw different colored circles
public class Bridge {
  public static void Main(string[] args){
  // implementation of concrete circles selected at run time.
  Shape redCircle = new Circle(100, 100, 10, new RedCircle());
  Shape greenCircle = new Circle(100, 100, 10, new GreenCircle());
  redCircle.Draw();
  greenCircle.Draw();
  }
}




